File: fontes/br.com.wdc.shopping/br.com.wdc.shopping.view.remote/remote.shell.teavm/src/main/java/br/com/wdc/shopping/view/remote/shell/teavm/bridge/FlushRequestContext.java
package br.com.wdc.shopping.view.remote.shell.teavm.bridge;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.teavm.jso.JSBody;
import org.teavm.jso.JSObject;

import br.com.wdc.shopping.view.teavm.commons.interop.Console;
import br.com.wdc.shopping.view.teavm.commons.interop.JsIntConsumer;
import br.com.wdc.shopping.view.teavm.commons.interop.JsRunnable;
import br.com.wdc.shopping.view.teavm.commons.interop.JsStringConsumer;
import br.com.wdc.shopping.view.teavm.commons.interop.Timers;

/**
 * Gerencia a conexão WebSocket, fila de requests, keepalive e estado de submitting.
 * Equivalente fiel ao FlushRequestContext.ts do remote.shell.react.
 */
public class FlushRequestContext {

    private static final int KEEP_ALIVE_INTERVAL = 15_000;

    private final ViewStateCoordinator app;

    private JSObject socket;
    private final Map<Integer, Map<String, Object>> requestMap = new LinkedHashMap<>();
    private final Set<Integer> userRequestIds = new HashSet<>();

    int lastSentRequestId = -1;
    int requestCount = 0;
    int lastProcessedId = -1;
    private int keepAliveHandler = 0;
    private int pendingKeepAlive = 0;
    private int submittingTimer = 0;
    private int submittingTimeout = 0;
    private String pendingSecret;

    public FlushRequestContext(ViewStateCoordinator app) {
        this.app = app;

        // Restore request counter from sessionStorage (survives F5)
        String savedSeq = getSessionItem("req_seq");
        if (savedSeq != null && !savedSeq.isEmpty()) {
            try {
                int parsed = Integer.parseInt(savedSeq);
                if (parsed > 0) {
                    this.requestCount = parsed;
                }
            } catch (NumberFormatException ignored) {
                // Intentionally empty: non-numeric session counter is safely ignored
            }
        }
    }

    public void submit(Map<String, Object> formMap, String vsid, int eventId, boolean silent) {
        cancelPendingKeepAlive();

        formMap.put("requestId", requestCount);
        @SuppressWarnings("unchecked")
        List<String> events = (List<String>) formMap.get("event");
        if (events == null) {
            events = new ArrayList<>();
            formMap.put("event", events);
        }
        events.add(vsid + ":" + eventId);

        requestMap.put(requestCount, formMap);
        if (!silent) {
            userRequestIds.add(requestCount);
        }
        requestCount++;
        persistState();

        resetKeepAliveTimer();
        flush();
    }

    @SuppressWarnings("unchecked")
    public void flush() {
        if (socket == null || !isSocketOpen(socket)) {
            return;
        }

        Map<String, Object> requestObj = new LinkedHashMap<>();
        List<String> allEvents = new ArrayList<>();
        boolean hasData = false;

        for (int i = lastSentRequestId + 1; i < requestCount; i++) {
            Map<String, Object> requestItemObj = requestMap.get(i);
            if (requestItemObj == null) {
                continue;
            }

            for (Map.Entry<String, Object> entry : requestItemObj.entrySet()) {
                String key = entry.getKey();
                Object value = entry.getValue();
                if (value == null) continue;

                if ("event".equals(key)) {
                    List<String> valueArray = (List<String>) value;
                    allEvents.addAll(valueArray);
                } else if ("requestId".equals(key)) {
                    // will be set below
                } else {
                    // Merge form data (nested map)
                    Map<String, Object> formData = (Map<String, Object>) requestObj.get(key);
                    if (formData == null) {
                        formData = new LinkedHashMap<>();
                        requestObj.put(key, formData);
                    }
                    if (value instanceof Map<?, ?> valueMap) {
                        formData.putAll((Map<String, Object>) valueMap);
                    } else {
                        requestObj.put(key, value);
                    }
                }
            }

            requestObj.put("requestId", i);
            lastSentRequestId = i;
            hasData = true;
        }

        if (hasData || pendingSecret != null) {
            requestObj.put("event", allEvents);
            if (pendingSecret != null) {
                requestObj.put("secret", pendingSecret);
                pendingSecret = null;
            }
            String json = JsonParser.stringify(requestObj);
            sendToSocket(socket, json);
            if (!userRequestIds.isEmpty()) {
                setSubmitting(true);
            }
        }
    }

    public void open(String url) {
        if (socket != null && isSocketOpenOrConnecting(socket)) {
            return;
        }

        Console.log("WebSocket opening: " + url);

        socket = createWebSocket(url);
        setupWebSocket(socket,
                (JsRunnable) this::handleOpen,
                (JsIntConsumer) this::handleClose,
                (JsRunnable) this::handleError,
                (JsStringConsumer) this::handleMessage);
    }

    public void close() {
        if (socket != null) {
            closeSocket(socket);
            socket = null;
        }
    }

    // -- WebSocket event handlers --

    private void handleOpen() {
        Console.log("WebSocket connected");
        app.isConnected = true;
        pendingSecret = getGlobalAppSignature();
        initKeepAliveChecks();
        flush();
    }

    private void handleClose(int code) {
        Console.warn("WebSocket closed: " + code);
        if (code == 4001) {
            // Session invalid: reload the page
            reload();
            return;
        }
        handleDisconnect("WebSocket closed: " + code);
    }

    private void handleError() {
        Console.error("WebSocket error");
        handleDisconnect("WebSocket error");
    }

    private void handleDisconnect(String cause) {
        socket = null;
        app.isConnected = false;
        stopKeepAliveChecks();
        userRequestIds.clear();
        setSubmitting(false);
        app.reconnectController.reconnect(cause);
    }

    @SuppressWarnings("unchecked")
    private void handleMessage(String data) {
        if (app.reconnectController.count > 0) {
            app.reconnectController.reset();
        }

        Map<String, Object> response = JsonParser.parseObject(data);
        if (response.isEmpty()) return;

        // GC: released views
        Object released = response.get("releasedViews");
        if (released instanceof List<?> releasedList) {
            app.viewGarbageCollector.release((List<String>) releasedList);
        }

        // GC: active views sweep
        Object active = response.get("activeViews");
        if (active instanceof List<?> activeList) {
            app.viewGarbageCollector.sweep((List<String>) activeList);
        }

        // Process requestId acknowledgment
        Object reqId = response.get("requestId");
        if (reqId instanceof Number n) {
            int processedId = n.intValue();
            for (int i = lastProcessedId + 1; i <= processedId; i++) {
                requestMap.remove(i);
                userRequestIds.remove(i);
                lastProcessedId = i;
            }
        }

        // Process URI change
        Object uri = response.get("uri");
        if (uri instanceof String u) {
            app.path = u;
            setLocationHref("#" + u);
        }

        // Process view states
        Object states = response.get("states");
        if (states instanceof List<?> list) {
            app.applyViewStates(list);
        }

        flush();
        setSubmitting(!userRequestIds.isEmpty());
    }

    // -- Keep alive --

    private void initKeepAliveChecks() {
        stopKeepAliveChecks();
        keepAliveHandler = Timers.setTimeout((Timers.TimerCallback) this::keepAlive, KEEP_ALIVE_INTERVAL);
    }

    private void stopKeepAliveChecks() {
        if (keepAliveHandler != 0) {
            Timers.clearTimeout(keepAliveHandler);
            keepAliveHandler = 0;
        }
    }

    private void keepAliveNow() {
        cancelPendingKeepAlive();
        if (socket != null && isSocketOpen(socket)) {
            pendingKeepAlive = Timers.setTimeout((Timers.TimerCallback) () -> {
                pendingKeepAlive = 0;
                persistState();
                sendToSocket(socket, "{\"ping\":true}");
            }, 80);
        }
    }

    private void cancelPendingKeepAlive() {
        if (pendingKeepAlive != 0) {
            Timers.clearTimeout(pendingKeepAlive);
            pendingKeepAlive = 0;
        }
    }

    private void resetKeepAliveTimer() {
        if (keepAliveHandler != 0) {
            stopKeepAliveChecks();
            keepAliveHandler = Timers.setTimeout((Timers.TimerCallback) this::keepAlive, KEEP_ALIVE_INTERVAL);
        }
    }

    private void keepAlive() {
        stopKeepAliveChecks();
        keepAliveNow();
        keepAliveHandler = Timers.setTimeout((Timers.TimerCallback) this::keepAlive, KEEP_ALIVE_INTERVAL);
    }

    // -- Submitting state --

    private void setSubmitting(boolean value) {
        if (value) {
            if (submittingTimer == 0) {
                submittingTimer = Timers.setTimeout((Timers.TimerCallback) () -> {
                    submittingTimer = 0;
                    applySubmitting(true);
                }, 200);
            }
            if (submittingTimeout == 0) {
                submittingTimeout = Timers.setTimeout((Timers.TimerCallback) () -> {
                    submittingTimeout = 0;
                    setSubmitting(false);
                }, 15_000);
            }
        } else {
            if (submittingTimer != 0) {
                Timers.clearTimeout(submittingTimer);
                submittingTimer = 0;
            }
            if (submittingTimeout != 0) {
                Timers.clearTimeout(submittingTimeout);
                submittingTimeout = 0;
            }
            applySubmitting(false);
        }
    }

    private void applySubmitting(boolean value) {
        var scope = app.viewMap.get(ViewStateCoordinator.BROWSER_VSID);
        if (scope != null) {
            var state = scope.getState();
            Object current = state.get("submitting");
            boolean currentVal = current instanceof Boolean b && b;
            if (currentVal != value) {
                state.put("submitting", value);
                scope.forceUpdate();
            }
        }
    }

    private void persistState() {
        setSessionItem("req_seq", String.valueOf(requestCount));
    }

    // -- JS interop --

    @JSBody(params = {"url"}, script = ""
            + "var s = new WebSocket(url, ['wdc']);"
            + "s.withCredentials = true;"
            + "return s;")
    private static native JSObject createWebSocket(String url);

    @JSBody(params = {"ws", "onOpen", "onClose", "onError", "onMessage"}, script = ""
            + "ws.onopen = onOpen;"
            + "ws.onclose = function(e) { onClose(e.code); };"
            + "ws.onerror = onError;"
            + "ws.onmessage = function(e) { onMessage(e.data); };")
    private static native void setupWebSocket(JSObject ws, JsRunnable onOpen, JsIntConsumer onClose, JsRunnable onError, JsStringConsumer onMessage);

    @JSBody(params = {"ws", "data"}, script = "ws.send(data);")
    private static native void sendToSocket(JSObject ws, String data);

    @JSBody(params = {"ws"}, script = "ws.close();")
    private static native void closeSocket(JSObject ws);

    @JSBody(params = {"ws"}, script = "return ws.readyState === 1;")
    private static native boolean isSocketOpen(JSObject ws);

    @JSBody(params = {"ws"}, script = "return ws.readyState === 0 || ws.readyState === 1;")
    private static native boolean isSocketOpenOrConnecting(JSObject ws);

    @JSBody(params = {}, script = "return window.__wdc_appSignature || null;")
    private static native String getGlobalAppSignature();

    @JSBody(params = {"key"}, script = "try { return sessionStorage.getItem(key); } catch(e) { return null; }")
    private static native String getSessionItem(String key);

    @JSBody(params = {"key", "val"}, script = "try { sessionStorage.setItem(key, val); } catch(e) {}")
    private static native void setSessionItem(String key, String val);

    @JSBody(params = {}, script = "location.reload();")
    private static native void reload();

    @JSBody(params = {"href"}, script = "window.location.href = href;")
    private static native void setLocationHref(String href);
}

File: fontes/br.com.wdc.framework/br.com.wdc.framework.cube.remote/remote.bridge.flutter/lib/src/bridge/flush_request_context.dart
import 'dart:async';
import 'dart:convert';

import 'package:web_socket_channel/web_socket_channel.dart';

import 'constants.dart';
import 'types.dart';
import 'view_state_coordinator.dart';

/// Manages the WebSocket connection and request queue.
///
/// Responsible for: flushing pending requests, keepalive pings,
/// processing server responses, and tracking the "submitting" state.
class FlushRequestContext {
  final ViewStateCoordinator app;

  WebSocketChannel? _channel;
  bool _isOpen = false;
  bool _isConnecting = false;
  final Map<int, FormMapType> _requestMap = {};
  int _lastSentRequestId = -1;
  int _requestCount = 0;
  int _lastProcessedId = -1;
  Timer? _keepAliveTimer;
  Timer? _pendingKeepAliveTimer;
  Timer? _submittingTimer;
  Timer? _submittingTimeout;
  final Set<int> _userRequestIds = {};
  String? _pendingSecret;
  String? _pendingAccessToken;

  FlushRequestContext(this.app) {
    // Restore request counter from platform storage (survives F5)
    final restored = app.onRestoreRequestSeq?.call() ?? 0;
    if (restored > 0) {
      _requestCount = restored;
    }
  }

  void submit(FormMapType formMap, String vsid, int eventId, {bool silent = false}) {
    _cancelPendingKeepAlive();

    formMap['requestId'] = _requestCount++;
    final events = formMap['event'] as List<String>?;
    if (events != null) {
      events.add('$vsid:$eventId');
    } else {
      formMap['event'] = <String>['$vsid:$eventId'];
    }
    _requestMap[formMap['requestId'] as int] = formMap;

    if (!silent) {
      _userRequestIds.add(formMap['requestId'] as int);
    }

    app.onPersistRequestSeq?.call(_requestCount);
    _resetKeepAliveTimer();
    flush();
  }

  void flush() {
    if (!_isOpen) return;

    final requestObj = <String, dynamic>{'event': <String>[]};
    var hasData = false;

    for (var i = _lastSentRequestId + 1; i < _requestCount; i++) {
      final requestItemObj = _requestMap[i];
      if (requestItemObj == null) continue;

      for (final entry in requestItemObj.entries) {
        final key = entry.key;
        final value = entry.value;
        if (value != null) {
          if (key == 'event') {
            (requestObj['event'] as List<String>).addAll(value as List<String>);
          } else if (value is Map) {
            final formData = requestObj.putIfAbsent(key, () => <String, dynamic>{}) as Map<String, dynamic>;
            formData.addAll(Map<String, dynamic>.from(value));
          } else {
            requestObj[key] = value;
          }
        }
      }
      requestObj['requestId'] = i;
      _lastSentRequestId = i;
      hasData = true;
    }

    if (hasData || _pendingSecret != null) {
      if (_pendingSecret != null) {
        requestObj['secret'] = _pendingSecret;
        _pendingSecret = null;
      }
      if (_pendingAccessToken != null) {
        requestObj['accessToken'] = app.dataSecurity.b64Cipher(_pendingAccessToken!);
        _pendingAccessToken = null;
      }
      _channel?.sink.add(jsonEncode(requestObj));
      if (_userRequestIds.isNotEmpty) {
        _startSubmitting();
      }
    }
  }

  void open(String url) {
    if (_isOpen || _isConnecting) return;
    _isConnecting = true;

    final channel = WebSocketChannel.connect(
      Uri.parse(url),
      protocols: ['wdc'],
    );
    _channel = channel;

    channel.ready.then((_) {
      _isConnecting = false;
      _isOpen = true;
      app.isConnected = true;
      _pendingSecret = app.dataSecurity.getSignature();
      _pendingAccessToken = app.accessToken;
      _initKeepAliveChecks();
      flush();
    }).catchError((Object error) {
      _isConnecting = false;
      _handleDisconnect(error);
    });

    channel.stream.listen(
      (dynamic message) {
        _onMessage(message as String);
      },
      onError: (Object error) {
        _handleDisconnect(error);
      },
      onDone: () {
        final closeCode = channel.closeCode;
        if (closeCode == 4001) {
          // Server says session invalid — notify app to reload
          app.onSessionInvalid();
          return;
        }
        _handleDisconnect('WebSocket closed (code: $closeCode)');
      },
    );
  }

  void close() {
    _isOpen = false;
    _isConnecting = false;
    _stopKeepAliveChecks();
    _cancelPendingKeepAlive();
    _stopSubmitting();
    _channel?.sink.close();
    _channel = null;
  }

  void _handleDisconnect(Object? cause) {
    if (_channel == null) return; // Intentional close — don't reconnect
    _isOpen = false;
    _channel = null;
    app.isConnected = false;
    _stopKeepAliveChecks();
    // Clean up sent-but-unacknowledged requests (won't be retried)
    for (var i = _lastProcessedId + 1; i <= _lastSentRequestId; i++) {
      _requestMap.remove(i);
    }
    _userRequestIds.clear();
    _stopSubmitting();
    app.reconnectController.reconnect(cause);
  }

  void _onMessage(String rawMessage) {
    if (app.reconnectController.count > 0) {
      app.reconnectController.reset();
    }

    final response = jsonDecode(rawMessage) as Map<String, dynamic>;

    if (response['releasedViews'] != null) {
      app.viewGarbageCollector.release(response['releasedViews'] as List<dynamic>);
    }
    if (response['activeViews'] != null) {
      app.viewGarbageCollector.sweep(response['activeViews'] as List<dynamic>);
    }

    if (response['requestId'] != null) {
      final responseRequestId = response['requestId'] as int;
      for (var i = _lastProcessedId + 1; i <= responseRequestId; i++) {
        _requestMap.remove(i);
        _userRequestIds.remove(i);
        _lastProcessedId = i;
      }
    }

    if (response['uri'] != null) {
      app.onUriChanged(response['uri'] as String);
    }

    if (response.containsKey('accessToken')) {
      final ciphered = response['accessToken'] as String?;
      if (ciphered != null && ciphered.isNotEmpty) {
        final token = app.dataSecurity.b64Decipher(ciphered);
        app.accessToken = token;
        app.onAccessTokenChanged?.call(token);
      } else {
        app.accessToken = null;
        app.onAccessTokenChanged?.call('');
      }
    }

    if (response['states'] != null) {
      app.applyViewStates(response['states'] as List<dynamic>);
    }

    flush();
    if (_userRequestIds.isNotEmpty) {
      _startSubmitting();
    } else {
      _stopSubmitting();
    }
  }

  // :: Keep-alive

  void _initKeepAliveChecks() {
    _stopKeepAliveChecks();
    _keepAliveTimer = Timer(keepAliveInterval, _keepAlive);
  }

  void _stopKeepAliveChecks() {
    _keepAliveTimer?.cancel();
    _keepAliveTimer = null;
  }

  void _keepAliveNow() {
    _cancelPendingKeepAlive();
    if (_isOpen) {
      _pendingKeepAliveTimer = Timer(const Duration(milliseconds: 80), () {
        _pendingKeepAliveTimer = null;
        if (_isOpen) {
          _channel?.sink.add(jsonEncode({'ping': true}));
        }
      });
    }
  }

  void _cancelPendingKeepAlive() {
    _pendingKeepAliveTimer?.cancel();
    _pendingKeepAliveTimer = null;
  }

  void _resetKeepAliveTimer() {
    if (_keepAliveTimer != null) {
      _stopKeepAliveChecks();
      _keepAliveTimer = Timer(keepAliveInterval, _keepAlive);
    }
  }

  void _keepAlive() {
    _stopKeepAliveChecks();
    _keepAliveNow();
    _keepAliveTimer = Timer(keepAliveInterval, _keepAlive);
  }

  // :: Submitting state

  void _startSubmitting() {
    _submittingTimer ??= Timer(const Duration(milliseconds: 200), () {
      _submittingTimer = null;
      _applySubmitting(true);
    });
    _submittingTimeout ??= Timer(const Duration(seconds: 5), () {
      _submittingTimeout = null;
      _stopSubmitting();
    });
  }

  void _stopSubmitting() {
    _submittingTimer?.cancel();
    _submittingTimer = null;
    _submittingTimeout?.cancel();
    _submittingTimeout = null;
    _applySubmitting(false);
  }

  void _applySubmitting(bool value) {
    final scope = app.viewMap[browserVsid];
    if (scope != null && scope.state['submitting'] != value) {
      scope.patchField('submitting', value);
    }
  }
}
